File: game_client/Assets/Scripts/ResScripts/NPResCommon/BuildingCondition/_ABasicBuildingCondition.cs
using System;
using ALPackage;
using Common.ConditionEnum;
using GOE.Condition;

namespace GOE
{
    public abstract class _ABasicBuildingCondition : _ITNPBasicCondition<EBuildingConditionType
#if NP_GAME
        , BusinessBuildingRefObj
#endif
        , NPVarInfo
    >
    {
        /// <summary>
        /// 条件类型
        /// </summary>
        public abstract EBuildingConditionType conditionType { get; }
        
        /// <summary>
        /// 判断条件是否满足
        /// </summary>
        /// <param name="_data"></param>
        /// <param name="_varVariableInfo"></param>
        /// <returns></returns>
        public abstract bool isEnable(BusinessBuildingRefObj _data, NPVarInfo _varVariableInfo);
        
        /**************
            * 将字符串转化为本对象
            **/
        public static _ABasicBuildingCondition readCondition(string _str)
        {
            ALStringReader stringReader = new ALStringReader(_str);
            string condTypeS = stringReader.readItem(':');

            if (null == condTypeS)
            {
                ALLog.Warning("空的条件字符串:" + _str);
                return null;
            }

            EBuildingConditionType condType = EBuildingConditionType.NONE;
            try
            {
                //读取类型枚举
                condType = (EBuildingConditionType) ALCommon.EnumParse(typeof(EBuildingConditionType), condTypeS, true);
            }
            catch (Exception)
            {
                Debug.LogError($"建筑条件配置错误！！不存在类型:{condTypeS}，str:{_str}");
            }

            if (condType == EBuildingConditionType.NONE)
            {
#if UNITY_EDITOR
                ALLog.Warning("错误的条件类型:" + _str);
#endif
                return null;
            }

            //创建对象
            return readCondition(condType, stringReader);
        }

        /********************
        * 从节点中读取相关信息
        */
        public static _ABasicBuildingCondition readCondition(EBuildingConditionType _conditionType, ALStringReader _reader)
        {
            switch (_conditionType)
            {
                case EBuildingConditionType.NONE:
                    return new BuildingConditionFalse();
                case EBuildingConditionType.CS_BUILDING_ID:
                    return BuildingCondition_CS_BUILDING_ID.readStr(_reader);
                case EBuildingConditionType.CS_SPEC_ATTR_TYPE:
                    return BuildingCondition_CS_SPEC_ATTR_TYPE.readStr(_reader);
                default:
#if UNITY_EDITOR
                    Debug.LogError($"找不到condition：{_conditionType}==={_reader.srcString}");
#endif
                    return new BuildingConditionFalse();
            }
        }
    }
}

File: game_server/GameRes/src/NPGameRes/GameObjs/CommonObj/BuildingCondition/_ABasicBuildingCondition.java
package NPGameRes.GameObjs.CommonObj.BuildingCondition;

import ALServerLog.ALServerLog;
import Common.ConditionEnum.EBuildingConditionType;
import NPCommon.CommonObj.NPStringReader;
import NPGameRes.GameObjs.CommonObj.BuildingCondition.ConditionObj.BuildingCondition_CS_BUILDING_ID;
import NPGameRes.GameObjs.CommonObj.BuildingCondition.ConditionObj.BuildingCondition_CS_SPEC_ATTR_TYPE;
import NPGameRes.GameObjs.CommonObj.BuildingCondition.ConditionObj.BuildingCondition_NONE;
import NPGameRes.GameObjs.CommonObj.Condition.InterfaceObj._ITNPBasicCondition;

abstract public class _ABasicBuildingCondition implements _ITNPBasicCondition<EBuildingConditionType>
{
    /**************
     * 将字符串转化为本对象
     **/
    public static _ABasicBuildingCondition readCondition(String _str)
    {
        NPStringReader stringReader = new NPStringReader(_str);
        //读取类型字符串
        String condTypeS = stringReader.readItem(':');

        if (null == condTypeS)
        {
            ALServerLog.Error("空的条件字符串:" + _str);
            return null;
        }

        EBuildingConditionType condType = EBuildingConditionType.NONE;
        try
        {
            //读取类型枚举
            condType = EBuildingConditionType.valueOf(condTypeS.toUpperCase());
        } catch (Exception e)
        {
        }

        if (condType == EBuildingConditionType.NONE)
        {
            ALServerLog.Error("错误的条件类型:" + _str);
            return null;
        }

        return readCondition(condType, stringReader);
    }

    /********************
     * 从节点中读取相关信息
     */
    public static _ABasicBuildingCondition readCondition(EBuildingConditionType _conditionType, NPStringReader _reader)
    {
        switch (_conditionType)
        {
            case NONE:
                return new BuildingCondition_NONE();
            case CS_BUILDING_ID:
                return BuildingCondition_CS_BUILDING_ID.readCond(_reader);
            case CS_SPEC_ATTR_TYPE:
                return BuildingCondition_CS_SPEC_ATTR_TYPE.readCond(_reader);
            default:
                return new BuildingCondition_NONE();
        }
    }
}
